### ISSUE TITLE
Image keyboard support

### ISSUE THREAD

**Issue by netdur** [flutter/flutter#20796]
It is possible to implement Commit Content API using Flutter's TextField or TextFormField or any widget?

https://developer.android.com/guide/topics/text/image-keyboard

I am not sure convert this line

layout.addView(createEditTextWithContentMimeTypes(new String[]{"image/gif"}));

from this code example

https://github.com/googlesamples/android-CommitContentSampleApp/blob/master/app/src/main/java/com/example/android/commitcontent/app/MainActivity.java

**Comment by mehmetf** [comment]
This came up internally as well as a question. Could we get a comment on the scope of this request? It feels like a fairly large one to be able to insert an image to Flutter via IME.

**Comment by audkar** [comment]
Is there any known workarounds? This basically prevents using flutter in any app which has chat functionality for younger audience...

**Comment by gerry05** [comment]
Still having a problem with this one. I can't send a gif/sticker. I'm working with a chat app.

**Comment by wilburt** [comment]
Please is there any update on this?

**Comment by RuedigerMoeller** [comment]
would be fine to at least have a custom callback for special content on TextField. Not necessary to implement full blown richtext editor ..

**Comment by animeshjain** [comment]
+1 for the chat use case. Most keyboards have some way to search for GIFs and chat is no fun without all those memes!

**Comment by robert-virkus** [comment]
I am also interested in a workaround. Not to be able to use images from the Android and iOS keyboards is a major constraint for any communication app.

**Comment by AxesandGrinds** [comment]
I'm designing a chat app. Not having this feature just makes me feel like a fool for even attempting to use flutter for this app. And I have already gotten really far, I've developed most of the app, >70% and I can't send gifs or stickers, this is in fact evil of the flutter team.

**Comment by QuentinSc** [comment]
+1 it's a use case Flutter must covered

**Comment by itssidhere** [comment]
+1 Please work on this , it's a must have thing in a chat app.

**Comment by frankyvij** [comment]
It's 2020 and flutter still expects TextField and TextFormField to be plain text. Haha. When should we expect an update on this? Or if its not on the roadmap, least you can do is tell us!

**Comment by iPLEOMAX** [comment]
What @audkar basically said. Just give us a callback for TextField to handle the insertion manually.

Extremely important for any chat app.

**Comment by altanozlu** [comment]
Still nothing...

**Comment by antonkrasov** [comment]
Ok, I made it work, the main problem here is that it would require updates for both engine and framework. I would probably try to work on it in the next couple of weeks, lots of work here, wait for PR guys :)

![ezgif-4-269db5a08874](https://user-images.githubusercontent.com/3696113/88133768-b5ded000-cb97-11ea-80f3-5b873f0ffd46.gif)

**Comment by kanily-tech** [comment]
@antonkrasov Whoa! Is it working for both iOS and Android?

**Comment by Bricktheworld** [comment]
That is super nice! Please keep us updated

**Comment by antonkrasov** [comment]
@kanily-tech ok, so spent lots of time today researching how to make it work on iOs and that's not so simple. Simple like for iOs the only way to do it is to use clipboard, and ask users to paste your content. Because I'm not an iOs guy, couldn't manage it. Most of the apps seems like using own GIFs providers, e.g. WhatsApp, Telegram, FB Messanger...

Guys, another news is that I won't need it anymore for myself, here are my findings, hopefully, someone will pick it up and finish: 

- https://github.com/antonkrasov/flutter/commit/0904d09b2311a9dfdf0c4330f03b365ccb8eb0db
- https://github.com/antonkrasov/engine/commit/acf26afac481ff59078531fbe18641defc32865d

**Comment by rahuldubey093** [comment]
It's a shame Flutter doesn't support a must needed feature. I have reached certain milestones in developing my chat app, even after facing roadblocks for petty things, and now when I am reaching towards completion, I realise I can't take GIF/Stickers input from user favourite keyboards like Gboard. Sure I have developed a custom keyboard with GIF and Stickers support but still no chat app can deny its users the right to also use their choice of keyboard. I can't go back from this stage, really hoping for an urgent solution for both Android and iOS.

**Comment by Aashikr** [comment]
I badly need the gif keyboard as well

**Comment by Bricktheworld** [comment]
So if you guys compile the engine using that repo, it has been working great for our team. It is basically a perfect solution, haven't encountered any bugs yet

**Comment by Bricktheworld** [comment]
However, I might add that this only works for android

**Comment by jimmyff** [comment]
It's great someone has taken the time to start work on this but using a custom build of the flutter engine is just not really an option for a production apps.

It would be great to get an update from the flutter team on 1st party support for Image Keyboard support. Hopefully they may be able to make use of the work that has already been done if it's inline with how they would implement the feature.

**Comment by lucasbiel7** [comment]
I'm looking for this feature

**Comment by FNPCMDs** [comment]
+1

**Comment by iPLEOMAX** [comment]
Is there a way to increase the priority on this? Its been too long.

**Comment by zlshames** [comment]
> Is there a way to increase the priority on this? Its been too long.

We've had it implemented in our app for quite some time, with no hiccups. Though, the installation process is a huge pain because you need to build your own flutter engine, then edit the flutter framework as well. Anytime there is an update to either, you have to re-merge the code from antonkrasov's repo.

It would be awesome to have someone write the required tests and what not for this so it can be implemented into both the framework and engine. That way we don't keep needing to rebuild each update, which can take hours.

**Comment by iPLEOMAX** [comment]
> > Is there a way to increase the priority on this? Its been too long.
> 
> We've had it implemented in our app for quite some time, with no hiccups. Though, the installation process is a huge pain because you need to build your own flutter engine, then edit the flutter framework as well. Anytime there is an update to either, you have to re-merge the code from antonkrasov's repo.
> 
> It would be awesome to have someone write the required tests and what not for this so it can be implemented into both the framework and engine. That way we don't keep needing to rebuild each update, which can take hours.

Alright I'm guessing there's no other choice right now. I'll look into building my own as you said.

How frequently do you usually have to rebuild for a production app? Generally speaking.

**Comment by zlshames** [comment]
> > > Is there a way to increase the priority on this? Its been too long.
> > 
> > 
> > We've had it implemented in our app for quite some time, with no hiccups. Though, the installation process is a huge pain because you need to build your own flutter engine, then edit the flutter framework as well. Anytime there is an update to either, you have to re-merge the code from antonkrasov's repo.
> > It would be awesome to have someone write the required tests and what not for this so it can be implemented into both the framework and engine. That way we don't keep needing to rebuild each update, which can take hours.
> 
> Alright I'm guessing there's no other choice right now. I'll look into building my own as you said.
> 
> How frequently do you usually have to rebuild for a production app? Generally speaking.

It really just depends. If there is a flutter framework update, fixing it to handle GIFs is fairly quick because you don't need to rebuild the whole engine. If there is an engine update, that's when it gets annoying because you'll have to pull the new stuff, merge it with the GIF keyboard code, then rebuild (which can take hours). Honestly, since setting up my engine a few months ago, I have not updated it. I haven't seen any issues. But I'm sure if I update my engine, there will be a fair amount of good fixes/optimizations that could benefit the app

**Comment by skpatrat** [comment]
> @kanily-tech ok, so spent lots of time today researching how to make it work on iOs and that's not so simple. Simple like for iOs the only way to do it is to use clipboard, and ask users to paste your content. Because I'm not an iOs guy, couldn't manage it. Most of the apps seems like using own GIFs providers, e.g. WhatsApp, Telegram, FB Messanger...
> 
> Guys, another news is that I won't need it anymore for myself, here are my findings, hopefully, someone will pick it up and finish:
> 
> * [antonkrasov@0904d09](https://github.com/antonkrasov/flutter/commit/0904d09b2311a9dfdf0c4330f03b365ccb8eb0db)
> * [antonkrasov/engine@acf26af](https://github.com/antonkrasov/engine/commit/acf26afac481ff59078531fbe18641defc32865d)

This was much needed. Thanks! 
Any way to merge this feature with flutter's main thread?

**Comment by jimmyff** [comment]
One way really to get this on the flutter teams radar is to add a thumbs up reaction to the first post. Then when issues are sorted by reactions (which is often looked at when prioritising tasks) - this issue will be up there, at the moment it doesn't even feature on the first page of results. 

If you haven't already, please add a reaction to the first post which will help give this issue visibility.

**Comment by brianschardt** [comment]
what about this package: https://pub.dev/packages/extended_text_field

**Comment by chinmaygarde** [comment]
cc @cbracken

**Comment by zlshames** [comment]
> https://pub.dev/packages/extended_text_field

Full disclosure, I have never tried this package for this purpose. However, I assume that this textfield will be able to display the data, but not get the data out of the keyboard. The way the google keyboard works is when you select a GIF, it basically sends the direct link (URL) for the GIF to the app that can handle it. Currently, flutter doesn't support receiving this link/data from the keyboard. Even with using the text field, you wouldn't be able to get the link; only display a link, if you had it.

In Anton Krasov's solution, he had to make changes both to Flutter's engine as well as the framework.

**Comment by themadmrj** [comment]
This feature is still very much missed and needed!

**Comment by emilekm2142** [comment]
Lack of this feature makes Flutter totally unusable for any chats. It is a shame that it does not work yet. Just provide us with a callback.
I am **very** disappointed.  

Ok, basically furious after discovering it does not work.

**Comment by duongtruong12** [comment]
please update this, this seem like a high property issue

**Comment by zlshames** [comment]
I've been compiling my own Flutter Engine to use this. However, with the release of v2.x, it makes it a bit harder to setup. Would be great to get this into the actual release so we don't need to compile our own engine

**Comment by ZackMitkin** [comment]
@zlshames 
Does it work well? I was thinking of doing the same

**Comment by zlshames** [comment]
> @zlshames
> Does it work well? I was thinking of doing the same

It works just OK. For context, I have it setup with v1.22.4. It works well with GBoard as you are just sending a URL. It _does not_ work with Samsung's GIF Keyboard, and it does not work with the Bitmoji keyboard "language" or whatever it is. I believe those are both app-types that send the actual image data as the content commit vs. sending just a URL to the GIF.

I did spend a few hours trying to add some fail-safes to Anton's code, but when I went to recompile the engine, it failed because of `gclient sync` trying to fetch the master branch of the upstream repo (main flutter engine repo), so it was overwriting my local changes to have GIF insertion. I think what I have to do is make the merge, and then submit a PR to my forked repo and get it merged into master of my own repo. Then I need to setup the `.gclient` file to point to my forked repository instead of the master repo.

**Comment by zlshames** [comment]
For all wondering, I was able to put together a proof of concept using Robert's proof of concept above, combined with Anton's proof of concept above. The result is Image Keyboard support for GBoard, Samsung Keyboard, Bitmoji, and more!

You can see how I implemented it in my engine repo here: https://github.com/BlueBubblesApp/engine/commits/master

It's all the commits starting with the `Image Keyboard Support` and newer. I had to make some smaller changes as I was testing the new engine. If you want to use this to build your own engine, you can do that. Just use my repo master branch with `gclient`. You'll also need to make the required changes in the Flutter Framework, but that's fairly easy, and you can modify it based on Anton's changes. 

I want to make an Ubuntu v18.04 build that others can use, but it will mean me uploading 15GB of data, and I don't know if I want to do that. I could also try to make a Docker container, but I gotta find time to do that

**Comment by darkfrog26** [comment]
So, Google, who created Gboard with a GIF insert feature, created Android, and created Flutter, doesn't support this? It's like if Tesla decided to create gas stations, but only provided gasoline...

**Comment by zim32** [comment]
I can not understand why this is still issue after three years. The only reasonable answer is google do not want you to build chat apps with flutter. He google team. What is the point if makine your own OS, your language, your framework if you can not solve issue more then three years old?

**Comment by tneotia** [comment]
@zlshames and I have been working on this the past week and have successfully implemented image keyboard support on Flutter 2.0+ engine and SDK.

We are currently finalizing things and hope to have a PR for the Android portion of image keyboard support out in the next few days. 

Neither of us are very experienced in iOS development (and our app doesn't even support iOS) so we currently do not have a solution for it.

**Comment by 0Vipin0** [comment]
After going through all the comments, I see that all the people who are using as feature in their app have created there own fork of the engine and framework itself and then used it compile the app. I just want to ask the flutter team if they have any priority for this in the future or in the roadmap as this seems to very basic feature for the chat app or any app that uses chat as a feature.

_@Hixie just yes or no or something might help in this regard?_ 

I love flutter by heart and I was unaware about this issue, and when I came across this issue I was little sad that this issue is not handled even after almost 3 years. This makes using chat as a feature in the app built using flutter very difficult.

**Comment by tneotia** [comment]
@0Vipin0 apologies for the delay. We are still working on our fork with image keyboard support for Android. We are trying to get a stable release out to our users to verify our solution works correctly before we make a PR to Flutter, that way we are certain there are no bugs or issues. Google Play has had an unusually long review process on our app and it is still pending. 

We will definitely submit a PR, I can say that for certain. Just when is down to when we feel confident it is stable and reliable enough to be merged into Flutter itself.

**Comment by zlshames** [comment]
I know that @tneotia already posted this link, but if you are wanting to use IME insertion _before_ the flutter team implements it, I have a created a tag in our repo.

https://github.com/BlueBubblesApp/flutter/tree/v2.2.3-ime-insertion

The tag is the v2.2.3 Flutter Framework + our custom IME insertion support. In order for this to work, you'll need to manually compile the dart engine (on Ubuntu 18.04). This engine code is the cherry picks from the corresponding Dart engine for Flutter 2.2.3 + our custom IME insertion support

https://github.com/BlueBubblesApp/engine

Luckily, the Flutter team has a pretty straightforward guide on this, found here: 

https://github.com/flutter/flutter/wiki/Compiling-the-engine#compiling-for-android-from-macos-or-linux

**Comment by 0Vipin0** [comment]
@tneotia @zlshames Thank you for your quick response.
> We will definitely submit a PR, I can say that for certain. Just when is down to when we feel confident it is stable and reliable enough to be merged into Flutter itself.

Completely make sense to me and that be very helpful to the flutter community itself.

Also Kudos to everyone for maintaining both codebase for flutter v1 and flutter v2. Since most large apps are yet to shift to the flutter v2 and having a specific version for flutter v1 helps a lot.

> One MAJOR issue in my opinion is that Google is moving to AppBundles instead of APKs for new apps after August 2021. When using a custom engine, there is no way to build an app bundle for multiple CPU architectures (i.e. arm, arm64, etc.). As a result, we can only use APKs if we want to support multiple architectures.

> If the Flutter team doesn't want to add this as a feature, they need to at minimum allow us to build an App bundle using multiple local engines so that we can support multiple architectures on the Play Store.

I am not aware of this issue, good to know this before hand and will check both repos posted above.

**Comment by Hixie** [comment]
If people would like to contribute implementations of this feature then we would certainly be happy to have it go through our normal [contribution process](https://github.com/flutter/flutter/blob/master/CONTRIBUTING.md). I would recommend starting with a design doc, so we can discuss the design (making sure it's appropriate for all the platforms we support); participating on our Discord is probably the best way to get involved. (This is the same process that anyone, including someone from Google, would go through when implementing this feature.)

As of right now I'm not aware of anyone having committed to implementing this (whether employed by Google or otherwise). There's a lot of low-level work going on right now with the keyboard support to enable proper support of keyboards across macOS, Linux, and Windows, which is probably higher priority, since getting the low level logic right will enable higher-level logic to work without churn.

**Comment by tneotia** [comment]
@Hixie we can definitely do that, though our implementation will only support Android. To be honest there is very minimal code & changes required to make this work, but if you'd prefer a design doc we can definitely make one. Thanks for your reply!

**Comment by Hixie** [comment]
If you have code already then I don't want to make you just do work for no reason, I recommend submitting PRs tentatively and seeing what the reaction is. If it's "oh yeah this is obvious, sounds good", then great. If it's "we should think about this more", then maybe then a design doc makes sense.

I recommend cc'ing @justinmc and @gspencergoog on the PRs.

**Comment by tneotia** [comment]
PRs created for both engine & SDK (linked above). Thanks everybody for waiting on these!

**Comment by netdur** [comment]
@tneotia you are god among mere mortals

**Comment by tneotia** [comment]
> @tneotia you are god among mere mortals

Thank @zlshames more than me ;)

Just here to help out as much as possible. This was enough of a blocker for us where we went out of our way to make a custom engine and custom SDK, but now we'd like to share this capability with the all other Flutter users! Someone's gotta do iOS though still....

**Comment by TariqHattab** [comment]
Hi, @tneotia and @zlshames !  
Thanks a lot for all the good work, I am currently working on Charma app a flutter social app with established 40k+ user base, and we would very much love to bring support of Bitmoji keyboard to our users, so I would like to know your opinion on whether I could or should use your solution to do that, if so are there any guides on how start? are there any concerns I should now about? 
Thank you very much.

**Comment by tneotia** [comment]
@TariqHattab if you wanted to start immediately your only option would be to clone our `engine` and `flutter` repo and use those when you are building/deploying. It is quite the hassle, but we have step-by-step instructions available [here](https://github.com/BlueBubblesApp/BlueBubbles-Android-App/wiki/Contribution-Guide#getting-gif-keyboard-support) if you want to go this route.

Then you can use the `onContentCommitted` callback we added to the `TextField` to get the image data for the inserted image content. The callback has dartdoc comments so you can use those to help you get started with the implementation in your app, otherwise message me personally and I can help you out.

I would recommend waiting for the PRs to be merged (hopefully they will be) just because the custom engine stuff is a hassle for production apps, but if it is a feature you need immediately then it can be used as a temporary workaround while the PRs are landed.

**Comment by zlshames** [comment]
> Hi, @tneotia and @zlshames !
> Thanks a lot for all the good work, I am currently working on Charma app a flutter social app with established 40k+ user base, and we would very much love to bring support of Bitmoji keyboard to our users, so I would like to know your opinion on whether I could or should use your solution to do that, if so are there any guides on how start? are there any concerns I should now about?
> Thank you very much.

Another downside to using our custom engine is you cannot build app-bundles for more than one CPU architecture. The `flutter build ...` command only allows specifying 1 CPU architecture at a time. This means you need to build the engine for each CPU architecture type you want to support. Then, run a separate command for each architecture to build the associated APK. The same goes for app-bundles. Right now we only build for arm and arm64 for that reason, and we build them separately.

**Comment by TariqHattab** [comment]
Hi, thank you for the reply, we well wait for an official fix hopefully soon, are there any updates or anyone working on bringing bitmoji keyboard  support in to flutter apps, 

Thank you very much

**Comment by kqureshi14** [comment]
@tneotia , can you please help me to figure out how to handle this response onContentCommitted against TextField as i followed your instructions and implemented now i am getting this response from TextField please advise how to display GIF from here.

CommittedContent(image/gif, content://com.google.android.inputmethod.latin.inputcontent/inputContent?fileName=%2Fdata%2Fuser%2F0%2Fcom.google.android.inputmethod.latin%2Ffiles%2Fgif55970583646172565620&packageName=com.tigerverde.mvp.chatApp&mimeType=image%2Fgif, [71, 73, 70, 56, 57, 97, 24, 1, 200, 0, 247, 255, 0, 1, 0, 1, 2, 0, 1, 2, 1, 2, 4, 1, 2, 4, 1, 4, 5, 3, 2, 5, 3, 4, 5, 3, 5, 6, 4, 6, 7, 4, 5, 8, 4, 5, 8, 6, 7, 10, 6, 6, 10, 7, 9, 10, 7, 10, 11, 4, 4, 11, 8, 11, 12, 9, 10, 13, 7, 6, 13, 10, 12, 14, 12, 15, 17, 14, 16, 17, 15, 18, 18, 7, 6, 18, 9, 10, 18, 12, 13, 19, 17, 19, 22, 17, 19, 23, 10, 10, 23, 16, 22, 23, 19, 20, 23, 21, 23, 24, 8, 5, 24, 13, 11, 24, 15, 15, 27, 22, 28, 28, 20, 19, 29, 10, 7, 29, 16, 14, 29, 17, 18, 29, 22, 23, 30, 14, 10, 30, 24, 24, 30, 25, 28, 33, 20, 18, 34, 27, 35, 35, 13, 9, 35, 17, 15, 35, 26, 28, 36, 29, 30, 38, 33, 36, 39, 20, 15, 39, 22, 18, 39, 23, 22, 40, 15, 11, 40, 26, 23, 42, 34, 41, 43, 29, 28, 44, 33, 33, 46, 35, 41, 47,

**Comment by zlshames** [comment]
> @tneotia , can you please help me to figure out how to handle this response onContentCommitted against TextField as i followed your instructions and implemented now i am getting this response from TextField please advise how to display GIF from here.
> 
> CommittedContent(image/gif, content://com.google.android.inputmethod.latin.inputcontent/inputContent?fileName=%2Fdata%2Fuser%2F0%2Fcom.google.android.inputmethod.latin%2Ffiles%2Fgif55970583646172565620&packageName=com.tigerverde.mvp.chatApp&mimeType=image%2Fgif, [71, 73, 70, 56, 57, 97, 24, 1, 200, 0, 247, 255, 0, 1, 0, 1, 2, 0, 1, 2, 1, 2, 4, 1, 2, 4, 1, 4, 5, 3, 2, 5, 3, 4, 5, 3, 5, 6, 4, 6, 7, 4, 5, 8, 4, 5, 8, 6, 7, 10, 6, 6, 10, 7, 9, 10, 7, 10, 11, 4, 4, 11, 8, 11, 12, 9, 10, 13, 7, 6, 13, 10, 12, 14, 12, 15, 17, 14, 16, 17, 15, 18, 18, 7, 6, 18, 9, 10, 18, 12, 13, 19, 17, 19, 22, 17, 19, 23, 10, 10, 23, 16, 22, 23, 19, 20, 23, 21, 23, 24, 8, 5, 24, 13, 11, 24, 15, 15, 27, 22, 28, 28, 20, 19, 29, 10, 7, 29, 16, 14, 29, 17, 18, 29, 22, 23, 30, 14, 10, 30, 24, 24, 30, 25, 28, 33, 20, 18, 34, 27, 35, 35, 13, 9, 35, 17, 15, 35, 26, 28, 36, 29, 30, 38, 33, 36, 39, 20, 15, 39, 22, 18, 39, 23, 22, 40, 15, 11, 40, 26, 23, 42, 34, 41, 43, 29, 28, 44, 33, 33, 46, 35, 41, 47,

Hey, the CommittedContent object has a uri that you should be able to read as a normal File I believe. Here is the example from our app:

```dart
void onContentCommit(CommittedContent content) async {
    // Add some debugging logs
    Logger.info("[Content Commit] Keyboard received content");
    Logger.info("  -> Content Type: ${content.mimeType}");
    Logger.info("  -> URI: ${content.uri}");
    Logger.info("  -> Content Length: ${content.hasData ? content.data!.length : "null"}");

    // Parse the filename from the URI and read the data as a List<int>
    String filename = uriToFilename(content.uri, content.mimeType);

    // Save the data to a location and add it to the file picker
    if (content.hasData) {
      this.addAttachments([PlatformFile(
        name: filename,
        size: content.data!.length,
        bytes: content.data,
      )]);

      // Update the state
      updateTextFieldAttachments();
      if (this.mounted) setState(() {});
    } else {
      showSnackbar('Insertion Failed', 'Attachment has no data!');
    }
  }
```

The uriToFilename function is here:

```dart
String uriToFilename(String? uri, String? mimeType) {
  // Handle any unknown cases
  String? ext = mimeType != null ? mimeType.split('/')[1] : null;
  ext = (ext != null && ext.contains('+')) ? ext.split('+')[0] : ext;
  if (uri == null) return (ext != null) ? 'unknown.$ext' : 'unknown';

  // Get the filename
  String filename = uri;
  if (filename.contains('/')) {
    filename = filename.split('/').last;
  }

  // Get the extension
  if (filename.contains('.')) {
    List<String> split = filename.split('.');
    ext = split[1];
    filename = split[0];
  }

  // Slugify the filename
  filename = slugify(filename, delimiter: '_');

  // Rebuild the filename
  return (ext != null && ext.length > 0) ? '$filename.$ext' : filename;
}
```

**Comment by kqureshi14** [comment]
@tneotia thanks for the direction but i need your further input as i followed your shared engine & flutter changes.

In your above comment, you used  this.addAttachments([PlatformFile()] is this your local function or added those in flutter or in Engine ?

**Comment by zlshames** [comment]
> @tneotia thanks for the direction but i need your further input as i followed your shared engine & flutter changes.
> 
> In your above comment, you used this.addAttachments([PlatformFile()] is this your local function or added those in flutter or in Engine ?

It is a custom function of ours. Unrelated to the flutter framework or engine

**Comment by kqureshi14** [comment]
@tneotia  thanks, i sorted out, it worked for me but ofcourse we will need the version from google to keep this working as whenever we will have Flutter update then we again have to face this  nightmare,  either this temp solution is for Android, if someone have IOS solution please share.

**Comment by kqureshi14** [comment]
@tneotia one question, in chat app although its working for me now in onCommittedChange and in that i handle via Firebase as i uploading the file there first and then get the URL to share in chat (in Stream) with Sender/Receiver, any advise on this to optimize this procedure ?

**Comment by tneotia** [comment]
> @tneotia one question, in chat app although its working for me now in onCommittedChange and in that i handle via Firebase as i uploading the file there first and then get the URL to share in chat (in Stream) with Sender/Receiver, any advise on this to optimize this procedure ?

While your image uploads to Firebase I recommend you use `Image.memory` and show the picture from bytes, then you can switch the widget to `Image.network` to show from network if you so wish. However I will advise against that - I suggest that in you chat app you download and cache all images locally and load them using `Image.file` or read as bytes and then use `Image.memory`. Main reason is that it will be easy for you to make sure users can view images even if the app is not connected to the internet.

**Comment by kqureshi14** [comment]
@tneotia  thanks for your great advise.

Plugin and Engine changes working fine in debug mode, but when i try to generate the release build i started get this error.
R8: Program type already present: io.flutter.embedding.engine.systemchannels.TextInputChannel$1

Is something i missed for release version ?

**Comment by tneotia** [comment]
> Is something i missed for release version ?

Not sure, haven't seen that error before.... seems like something is getting minified wrong, but I can't say for sure.

**Comment by kqureshi14** [comment]
@tneotia , actually i run your solution in Window 10, by just overwrite three files in "TextInputChannel.java" , "InputConnectionAdaptor.java", "TextInputPlugin.java" in Andriod Studio along with other "Flutter" framework level changes , but this solution worked for me in debug mode.
So now i am facing problem to make release build and facing this error

 R8: Program type already present: io.flutter.embedding.engine.systemchannels.TextInputChannel$1

so in order to make a release build,  apparently I have to download the complete engine as per your shared link but that will not work in Window 10, as you mentioned there  that will only work in Ubuntu 18.04,

Any other advise ?

**Comment by tneotia** [comment]
Yes, you need a Linux or macOS system to compile the engine and use it on release mode. Unfortunately Windows cannot compile so you would be out of luck to make a release build.

**Comment by fennelhans** [comment]
Really looking forward to this.

**Comment by zanderso** [comment]
From new feature request scan: @justinmc @GaryQian What is the status of this issue? What is left after https://github.com/flutter/engine/pull/35619?

**Comment by justinmc** [comment]
The framework PR (https://github.com/flutter/flutter/pull/110052) is all that remains, and I owe it a review.  Will post one today or tomorrow.

**Comment by ash-hashtag** [comment]
Really looking forward this

**Comment by zlshames** [comment]
Just an update, with Flutter 3.7, the GIF insertion code is now in the stable engine! All that's left is the merge into the flutter SDK repo here. But this also makes it much easier to manually implement the GIF insertion stuff on the SDK side of things (at least until the PR is merged). No more need to build the engine yourself.

I'll share our merged 3.7 SDK code with GIF insertion here when we've completed the merge. Should be shortly.

**Comment by zlshames** [comment]
Here is the SDK with the GIF insertion code merged into it: https://github.com/BlueBubblesApp/flutter

To use it just follow these instructions:
1. `cd` into your flutter repo code on your host (`~/snap/flutter/common` if you're on Linux)
2. Run: `git remote add upstream git@github.com:BlueBubblesApp/flutter.git`
3. Run: `git pull upstream master`

And that should be it. Now you can use the custom GIF insertion code with your text fields by adding a new parameter, like so:

```
contentInsertionConfiguration: ContentInsertionConfiguration(onContentInserted: (KeyboardInsertedContent content) {
    # ... your code handling the content
})
```

**Comment by zlshames** [comment]
FYI to everyone, this has been merged into Flutter! Shout out to @tneotia for getting it done! https://github.com/flutter/flutter/pull/110052#event-8403198915

**Comment by haquex19** [comment]
Was really looking forward to this, nice to know it is now supported. Is the guidance to just use the master channel to use this feature for now until it is in the stable channel?

**Comment by justinmc** [comment]
Yes, it's in [3.8.0-7.0.pre](https://github.com/flutter/flutter/releases/tag/3.8.0-7.0.pre) which is currently only available on master.

**Comment by vito-go** [comment]
> 3\. git pull upstream master

very very thank you. It worked. You really did me a favor.
